### ISSUE TITLE
Support Content-Encoding: gzip in URLDownloader

### ISSUE THREAD

**Issue by jessepeterson** [autopkg/autopkg#126]
As mentioned in autopkg/jessepeterson-recipes#5 I have a package download that is getting served with a `Content-Encoding: gzip`. It appears urllib2 doesn't natively support this. So: would folks be averse to merging a PR to implement that functionality in autopkg? I could, I suppose, also implement logic on the recipe and/or recipe Provider side to fixup the download.

The place where this is an actual issue is when the Unarchiver processor tries to run ditto to decompress a file labelled `.zip` but is actually encoded as a `.gz`.

**Comment by jessepeterson** [comment]
As a short term workaround how does this sound: Making the URLDownloader merely set a Content-Encoding output variable based on the corresponding HTTP header? In this way another recipe Processor step could take decisive action (or no action) based on what the content is encoded as.

I bring this up because the original recipe I had to work around this with *reverted* back to being non-gzip encoded. I'd guess perhaps they have differently configured load balancers or something: one that compresses and one that doesn't. Might be a crap shoot on what the download comes as so dynamically handling this would be valuable.

**Comment by timsutton** [comment]
Since we've got all the headers from the response already, we could just shove _all_ of them in an output variable dict like `response_headers`. Then if for some edge cases like yours, you have data to for a custom processor to make a decision about.

Having played around a bit with that, though:

```patch
diff --git a/Code/autopkglib/URLDownloader.py b/Code/autopkglib/URLDownloader.py
index 3aefe39..dad1fa8 100755
--- a/Code/autopkglib/URLDownloader.py
+++ b/Code/autopkglib/URLDownloader.py
@@ -90,6 +90,11 @@ class URLDownloader(Processor):
                 ("Boolean indicating if the download has changed since the "
                  "last time it was downloaded."),
         },
+        "response_headers": {
+            "description":
+                ("A dictionary containing all headers from the server "
+                 "response, keyed by header name."),
+        },
     }

     def main(self):
@@ -172,6 +177,7 @@ class URLDownloader(Processor):
                     self.output("Using existing %s" % pathname)
                     return

+            self.env["response_headers"] = url_handle.info().headers
             # Download file.
             self.env["download_changed"] = True
             with open(pathname, "wb") as file_handle:
```

...you wouldn't be guaranteed the output variable exists - the processor will exit early if it's deemed unmodified. Even moving up to the initial try: catch: (looking for a 304 response) wouldn't get you access to any useful data in the `url_handle`.

Maybe a better approach, then, would be for URLDownloader to properly handle gzip-encoded data transparently if needed. I'm kind of surprised that this hasn't been an issue before.

**Comment by jessepeterson** [comment]
Agreed: the *right* approach, I feel, is for URLDownloader to properly decode encoded data.

**Comment by gregneagle** [comment]
Also agreed.

**Comment by gregneagle** [comment]
"I'm kind of surprised that this hasn't been an issue before."

Web servers are not supposed to send gzip-deflated data in a response unless the client says it's OK in an "Accept-Encodings" header.

But since the clients here are sometimes internal software implementations inside autoupdated code, it's possible that some vendors may not worry about the client's preferences since they control the only client they care about.

**Comment by jessepeterson** [comment]
I guess if we hadn't just agreed on an approach and in that case it might have been exception-raising to have an encoded download that URLDownloader might've thrown.

**Comment by timsutton** [comment]
Ah, I see. And by default, urllib2 is setting the default `Accept-Encoding: identity`.

**Comment by jessepeterson** [comment]
I believe this issue is resolved now with the merging of PR #184. Thanks everyone!
